WalletWasabi/WabiSabi/Backend/Rounds/Arena.Partial.cs

```diff
@@ -329,4 +358,15 @@ public async Task<RoundState[]> GetStatusAsync(CancellationToken cancellationTok
 	private Alice GetAlice(Guid aliceId, Round round) =>
 		round.Alices.Find(x => x.Id == aliceId)
 		?? throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceNotFound, $"Round ({round.Id}): Alice ({aliceId}) not found.");
+
+	private static bool IsUserCheating(Exception e) =>
+		e is WabiSabiCryptoException || e is WabiSabiProtocolException wpe && wpe.ErrorCode
+			is WabiSabiProtocolErrorCode.InputSpent
```

Comment:
So I guess that one cheats when an intentionally bad data are being sent with requests but it is not cheating when, for example, network is being funny and a request is sent actually twice because it failed to be sent first time.

It seems to me that the enum values should be marked as such, so maybe we can modify `WabiSabiProtocolErrorCode.cs` and [add](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-create-a-new-method-for-an-enumeration#example) this list of error codes there. The advantage would be that modifying `WabiSabiProtocolErrorCode` would make it clear that cheating list must be updated too.

I don't know specifics but why `InputBanned` and `NonUniqueInputs` are not considered to be cheating?

Reply:
Many things here. Duplicated requests should be caught by the idempotent mechanism what should respond with the cached response.

The idea of moving this to the `WabiSabiProtocolErrorCode.cs` and add an extension method makes sense. Will do it.

`NonUniqueInputs` should be banned too because it is cheating, yes, good catch. `InputBanned` is not trying to cheat because the clients many/most of the times do not know they have been banned and re-ban them makes no difference.